packages/@aws-cdk/tmp-toolkit-helpers/src/api/io/private/action-aware.ts

```diff
@@ -28,3 +14,24 @@ export interface ActionAwareIoHost extends IIoHost {
   notify<T>(msg: ActionLessMessage<T>): Promise<void>;
   requestResponse<T, U>(msg: ActionLessRequest<T, U>): Promise<U>;
 }
+
+/**
+ * An IoHost wrapper that adds the given action to an actionless message before
+ * sending the message to the given IoHost
+ */
+export function withAction(ioHost: IIoHost, action: ToolkitAction): ActionAwareIoHost {
```

Comment:
Seems like we already have this function here:

https://github.com/aws/aws-cdk-cli/blob/01acccf590c67d10360ac46990958252a0663e44/packages/%40aws-cdk/toolkit-lib/lib/api/io/private/logger.ts#L14

Can we reuse?